File: interpretado/filaArray.py
class FilaArray:
    def __init__(self, capacity):  #construtr simples
        self.capacity = capacity
        self.arr = [None] * capacity
        self.front = 0
        self.rear = -1
        self.size = 0

    def isEmpty(self):
        return (self.size == 0)  #retorna true caso o tamanho da fila seja nulo
    
    def isFull(self):
        return (self.size == (self.capacity - 1))  #retorna true caso a fila tenha estourado a capacidade
    
    def enqueue(self, n):
        if self.isFull():
            print("erro! fila cheia..")
            return
        
        self.rear = (self.rear + 1) % self.capacity  #conserva as posicoes da fila de forma ciclica
        self.arr[self.rear] = n
        self.size += 1

    def dequeue(self):
        if self.isEmpty():
            print("erro! fila vazia..")
            return
        valor = self.arr[self.rear]
        self.rear = (self.rear + 1) % self.capacity
        self.size -= 1
        return valor  #retorna o valor retirado da fila
    
    def len(self):
        return self.size  #retorna o tamanho

File: compilado/src/filaArray.c
#include "../includes/libs.h"
#include "../includes/filaArray.h"


void criarFilaArray(FilaArray* fila) {  //instancia de criacao de uma fila vazia
    fila->front = 0;
    fila->rear = -1;
    fila->size = 0;
}

bool isEmptyFilaArray(FilaArray* fila) {
    return (fila->size == 0);  //verificacao simples se a fila esta vazia
}

bool isFullFilaArray(FilaArray* fila) {
    return (fila->size == MAX_SIZE); //verificacao simples se a fila esta cheia
}

void enqueueFilaArray(FilaArray* fila, int valor) {
    if (isFullFilaArray(fila)) {
        printf("erro! fila cheia..");  //verifica se a fila esta cheia
    }
    else{
    fila->rear = (fila->rear + 1) % MAX_SIZE; //o resto da divisao mantem a fila circular pois MAX_SIZE & MAX_SIZE = 0
    fila->arr[fila->rear] = valor;  //o valor se adiciona na fila
    
    fila->size++;  //o tamanho da fila aumenta
    }
}

int dequeueFilaArray(FilaArray* fila) {
    if (isEmptyFilaArray(fila)){
        printf("erro! fila vazia..");  //verifica se a fila esta vazia
        return -1;
    }

    int elementoRemovido = fila->arr[fila->front];  //aloca o elemento que sera removido na variavel local
    fila->front = (fila->front + 1) % MAX_SIZE;   //move a fila pra frente de maneira circular
    fila->size--;  //diminui o tamanho da fila
    
    return elementoRemovido;  //retorna o elemento removido
    
}

int lenFilaArray(FilaArray* fila) {
    return fila->size; //operacao simples de retornar tamanho
}

